iptables module

scootersmk opened the pull request:
Addresses one of the few features I currently use in serverspec that testinfra lacks. Open to suggestions on code. Thoughts?

philpep commented on testinfra/modules/iptables.py:
```diff
@@ -0,0 +1,50 @@
+        >>> host.file("/nonexistent").exists
+        False
+
+        """
```
docstring should be updated here.

Also the autodoc could be added in doc/source/modules.rst

scootersmk replied:
fixed

philpep commented on testinfra/modules/iptables.py:
```diff
@@ -0,0 +1,50 @@
+        return self.rule in rules
+
+    def get_rules(self):
+        rules = []
```
I'd be nice to have `get_rule(self, chain=None)` returning rules for a specific chain, for instance `host.iptable_rule().get_rules('FORWARD')` would return the output of `iptables -S FORWARD`

Also consider adding docstring with example of `get_rule` usages.

scootersmk replied:
fixed, let me know if you had something else in mind. See explanation in general discussion.

scootersmk commented:
Rewrote things a bit to handle specifying a chain. Here is my thinking on the two ways this can be used(demonstrated by the two different test blocks:
  1. Check for existence of a specific rule, optionally specifying a table. Since the rule contains the chain at the beginning, I thought passing 'chain' to get_rule here seemed redundant and made the arguments overly complex. 
  2. Pull a list of all rules that optionally match a specific table and/or chain. Then the user can perform as many tests on that list as they wish.

Updated docs/formatting to make all tests happy. I'm ready for this to be merged if you are.

philpep commented:
Sorry to not have spoted in my first review, but I think this module could have a better API.

What do you think of this api instead ?

```python
rules = host.iptables.rules(table='filter', chain='INPUT')
assert '-A INPUT -p tcp --dport http -j ACCEPT' in rules
assert '-A INPUT -j LOG' in rules
```

Where `rules(table=None, chain=None)` would simply return iptables rules as a python list.
You can inherit from `testinfra.modules.InstanceModule` to have `host.iptables` resolve to an instance of `Iptables` directly.

I find it's more simple (a single method `rules()` is needed), more pythonic (use of `in` operator) and less potential performance issues than with the `exists()` API (since each call will call a new `get_rules()`).

Tell me if you are ok with this API and if you have time to implement it (I you have no time, I can implement it based on your work).

Thanks, that's actually a big contribution!

philpep commented on testinfra/modules/iptables.py:
```diff
@@ -0,0 +1,87 @@
+        return self.rule in rules
+
+    def get_rules(self, table='filter', chain=None):
+        """Returns list of iptables rules by running
```
Looks like the `by running` is the beginning of an unfinished sentence, I'm wrong ?

scootersmk replied:
fixed

scootersmk commented:
Yes I agree with that API design as its both simple and flexible. Take a look at the most recent commit.

philpep commented:
Awesome, merged, thanks!
